fastestimator/network.py

```diff
@@ -13,51 +13,141 @@
 # limitations under the License.
 # ==============================================================================
 from collections import ChainMap
-from typing import Union, List, Dict, Any, Set, Mapping
+from typing import Any, Dict, List, Mapping, Set, Union
 
 import tensorflow as tf
 import torch
 
-from fastestimator.op import get_inputs_by_op, get_ops_by_mode, write_outputs_by_key, TensorOp
+from fastestimator.op import TensorOp, get_current_ops, get_inputs_by_op, write_outputs_by_key
 from fastestimator.op.tensorop.model import ModelOp, UpdateOp
-from fastestimator.util.util import NonContext, to_list
+from fastestimator.schedule import EpochScheduler, RepeatScheduler, Scheduler
+from fastestimator.util.util import NonContext, lcms, to_list
 
 
-class Network:
-    """A class representing network operations for FastEstimator model training.
-    Args:
-        ops : Specifies the series of operations for training model
-    """
-    def __init__(self, ops: Union[TensorOp, List[TensorOp]]):
+class BaseNetwork:
+    def __init__(self, ops):
         self.ops = to_list(ops)
-        self.framework = None
-        self.device = None
-        self.model_list = []
-        self.exported_keys = set()
-        self.op_outputs = set()
-        self.op_inputs = set()
-        self._initial_check()
-
-    def _initial_check(self):
-        self._check_model()
-        self._check_ops()
-        self._check_device()
-
-    def _check_ops(self):
+        self._verify_inputs()
+        self.effective_inputs = dict()
+        self.effective_outputs = dict()
+        self.epoch_ops = []
+
+    def _verify_inputs(self):
         for op in self.ops:
-            self.op_outputs = self.op_outputs.union(set(filter(None, to_list(op.outputs))))
-            self.op_inputs = self.op_inputs.union(set(filter(None, to_list(op.inputs))))
+            if isinstance(op, (RepeatScheduler, EpochScheduler)):
+                for epoch_op in op.get_items():
+                    assert isinstance(epoch_op, TensorOp), "unsupported op format, must provide TensorOp in Network"
+            else:
+                assert isinstance(op, TensorOp), "unsupported op format, must provide TensorOp in Network"
 
-    def _check_device(self):
-        if self.framework == "pytorch":
-            self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
-            if self.device.type == "cuda":
-                for model in self.model_list:
-                    model.to(self.device)
+    def load_epoch(self, mode, epoch):
+        self.epoch_ops = get_current_ops(self.ops, mode, epoch)
 
-    def prepare(self):
+    def unload_epoch(self):
         pass
 
+    def get_loss_keys(self) -> Set[str]:
+        loss_keys = set()
+        for op in self.ops:
+            if isinstance(op, (RepeatScheduler, EpochScheduler)):
```

Comment:
no functional scheduler here?

Reply:
`FunctionalScheduler` is designed for returning numerical values only. it will be used for learning rate scheduler for example.

Reply:
interesting. I think there's a way we can indicate that with type hinting, though I suppose that can wait